[libc++] Add FreeBSD-specific EINTEGRITY errc value

DimitryAndric opened the pull request:
Since ~2019 we have been carrying a libc++ patch in FreeBSD, that adds the `EINTEGRITY` errno value. At the time, I submitted https://reviews.llvm.org/D56398 for this, but due to the complicated `ELAST` handling in `errno.h`, which that patch made even more complicated, it was put on hold.

I would like to try this again, but much simplified: there is no need to modify `errno.h`, since all supported versions of FreeBSD define both `EOWNERDEAD` and `ENOTRECOVERABLE`, so the part that attempts to redefine `ELAST` is effectively never used, and for FreeBSD we can simply ignore it.

The only proposed changes will now be to optionally add `EINTEGRITY` as an `enum errc` value, and add a corresponding test to `errc.pass.cpp`.

DimitryAndric committed:
[libc++] Add FreeBSD-specific EINTEGRITY errc value

Since ~2019 we have been carrying a libc++ patch in FreeBSD, that adds
the `EINTEGRITY` errno value. At the time, I submitted
https://reviews.llvm.org/D56398 for this, but due to the complicated
`ELAST` handling in `errno.h`, which that patch made even more
complicated, it was put on hold.

I would like to try this again, but much simplified: there is no need to
modify `errno.h`, since all supported versions of FreeBSD define both
`EOWNERDEAD` and `ENOTRECOVERABLE`, so the part that attempts to
redefine `ELAST` is effectively never used, and for FreeBSD we can
simply ignore it.

The only proposed changes will now be to optionally add `EINTEGRITY` as
an `enum errc` value, and add a corresponding test to `errc.pass.cpp`.

DimitryAndric commented:
Note that reviews.llvm.org seems to have dropped off the internet, so for reference the old diff we are still carrying (which I want to get rid of) is the following:

```diff
--- a/libcxx/include/__system_error/errc.h
+++ b/libcxx/include/__system_error/errc.h
@@ -45,6 +45,7 @@ enum class errc
     identifier_removed,                 // EIDRM
     illegal_byte_sequence,              // EILSEQ
     inappropriate_io_control_operation, // ENOTTY
+    integrity_check_failed,             // EINTEGRITY
     interrupted,                        // EINTR
     invalid_argument,                   // EINVAL
     invalid_seek,                       // ESPIPE
@@ -142,6 +143,7 @@ _LIBCPP_DECLARE_STRONG_ENUM(errc)
     identifier_removed                  = EIDRM,
     illegal_byte_sequence               = EILSEQ,
     inappropriate_io_control_operation  = ENOTTY,
+    integrity_check_failed              = EINTEGRITY,
     interrupted                         = EINTR,
     invalid_argument                    = EINVAL,
     invalid_seek                        = ESPIPE,
--- a/libcxx/include/errno.h
+++ b/libcxx/include/errno.h
@@ -34,49 +34,79 @@ Macros:
 
 #ifdef __cplusplus
 
-#if !defined(EOWNERDEAD) || !defined(ENOTRECOVERABLE)
+#if !defined(EOWNERDEAD) || !defined(ENOTRECOVERABLE) || !defined(EINTEGRITY)
 
 #ifdef ELAST
 
 static const int __elast1 = ELAST+1;
 static const int __elast2 = ELAST+2;
+static const int __elast3 = ELAST+3;
 
 #else
 
 static const int __elast1 = 104;
 static const int __elast2 = 105;
+static const int __elast3 = 106;
 
 #endif
 
-#ifdef ENOTRECOVERABLE
+#if !defined(EOWNERDEAD) && !defined(ENOTRECOVERABLE) && !defined(EINTEGRITY)
+#define ENOTRECOVERABLE __elast1
+#define EOWNERDEAD __elast2
+#define EINTEGRITY __elast3
+#if defined(ELAST)
+#undef ELAST
+#define ELAST EINTEGRITY
+#endif
 
+#elif !defined(EOWNERDEAD) && !defined(ENOTRECOVERABLE) && defined(EINTEGRITY)
+#define ENOTRECOVERABLE __elast1
+#define EOWNERDEAD __elast2
+#if defined(ELAST)
+#undef ELAST
+#define ELAST EOWNERDEAD
+#endif
+
+#elif !defined(EOWNERDEAD) && defined(ENOTRECOVERABLE) && !defined(EINTEGRITY)
 #define EOWNERDEAD __elast1
+#define EINTEGRITY __elast2
+#if defined(ELAST)
+#undef ELAST
+#define ELAST EINTEGRITY
+#endif
 
-#ifdef ELAST
+#elif !defined(EOWNERDEAD) && defined(ENOTRECOVERABLE) && defined(EINTEGRITY)
+#define EOWNERDEAD __elast1
+#if defined(ELAST)
 #undef ELAST
 #define ELAST EOWNERDEAD
 #endif
 
-#elif defined(EOWNERDEAD)
+#elif defined(EOWNERDEAD) && !defined(ENOTRECOVERABLE) && !defined(EINTEGRITY)
+#define ENOTRECOVERABLE __elast1
+#define EINTEGRITY __elast2
+#if defined(ELAST)
+#undef ELAST
+#define ELAST EINTEGRITY
+#endif
 
+#elif defined(EOWNERDEAD) && !defined(ENOTRECOVERABLE) && defined(EINTEGRITY)
 #define ENOTRECOVERABLE __elast1
-#ifdef ELAST
+#if defined(ELAST)
 #undef ELAST
 #define ELAST ENOTRECOVERABLE
 #endif
 
-#else  // defined(EOWNERDEAD)
-
-#define EOWNERDEAD __elast1
-#define ENOTRECOVERABLE __elast2
-#ifdef ELAST
+#elif defined(EOWNERDEAD) && defined(ENOTRECOVERABLE) && !defined(EINTEGRITY)
+#define EINTEGRITY __elast1
+#if defined(ELAST)
 #undef ELAST
-#define ELAST ENOTRECOVERABLE
+#define ELAST EINTEGRITY
 #endif
 
-#endif // defined(EOWNERDEAD)
+#endif // !defined(OWNERDEAD) && !defined(NOTRECOVERABLE) && !defined(INTEGRITY)
 
-#endif // !defined(EOWNERDEAD) || !defined(ENOTRECOVERABLE)
+#endif // !defined(OWNERDEAD) || !defined(NOTRECOVERABLE) || !defined(INTEGRITY)
 
 //  supply errno values likely to be missing, particularly on Windows
 
@@ -392,6 +422,10 @@ static const int __elast2 = 105;
 
 #ifndef EMLINK
 #define EMLINK 9979
+#endif
+
+#ifndef EINTEGRITY
+#define EINTEGRITY 9980
 #endif
 
 #endif // __cplusplus
```

DimitryAndric commented on libcxx/include/__system_error/errc.h:
```diff
@@ -141,19 +142,22 @@ _LIBCPP_DECLARE_STRONG_ENUM(errc){
+    no_buffer_space     = ENOBUFS,
+    no_child_process    = ECHILD,
+    no_link             = ENOLINK,
+    no_lock_available   = ENOLCK,
```
Note that these changes were the result of clang-format.

philnik777 commented on libcxx/include/__system_error/errc.h:
```diff
@@ -141,19 +142,22 @@ _LIBCPP_DECLARE_STRONG_ENUM(errc){
-    no_link                            = ENOLINK,
-    no_lock_available                  = ENOLCK,
+#ifdef EINTEGRITY
+    integrity_check_failed = EINTEGRITY,
```
We should also add a comment that this is an extension.

philnik777 reviewed (changes requested):
Personally I'm fine with this specific extension, but I'd like to hear from @ldionne whether he is fine with this too. There should also be a new entry for https://libcxx.llvm.org/UsingLibcxx.html#libc-extensions. Maybe add something like "platform-specific extensions"?

mordante commented on libcxx/include/__system_error/errc.h:
```diff
@@ -141,19 +142,22 @@ _LIBCPP_DECLARE_STRONG_ENUM(errc){
-    no_child_process                   = ECHILD,
-    no_link                            = ENOLINK,
-    no_lock_available                  = ENOLCK,
+#ifdef EINTEGRITY
```
+1. Also we tend to document our extensions. Can you add this to https://libcxx.llvm.org/UsingLibcxx.html#libc-extensions

DimitryAndric commented:
Hm, I wonder which errno values are actually standardized? I mean, all the Linux-specific errno values in this header are also 'extensions' in some way?

DimitryAndric commented:
Oh I see now, the `enum class errc` at the top of `libcxx/include/__system_error/errc.h` is precisely the same list as in [system.error.syn]. I guess it makes sense to put extensions at the end of these lists? Or do we prefer alphabetical order? (I would think so, otherwise looking up a specific value is cumbersome.)

DimitryAndric committed:
Mark EINTEGRITY value as FreeBSD extension, and add subsection
"Extensions to <system_error>" to the "Libc++ Extensions" section in the
libcxx documentation.

DimitryAndric commented:
Hm, it looks like clang-format went completely crazy due to my latest change. For some reason it now wants to reformat the enum as:

```c++
_LIBCPP_DECLARE_STRONG_ENUM(errc) {
  address_family_not_supported = EAFNOSUPPORT, address_in_use = EADDRINUSE, address_not_available = EADDRNOTAVAIL,
  already_connected = EISCONN, argument_list_too_long = E2BIG, argument_out_of_domain = EDOM, bad_address = EFAULT,
  bad_file_descriptor = EBADF, bad_message = EBADMSG, broken_pipe = EPIPE, connection_aborted = ECONNABORTED,
  connection_already_in_progress = EALREADY, connection_refused = ECONNREFUSED, connection_reset = ECONNRESET,
  cross_device_link = EXDEV, destination_address_required = EDESTADDRREQ, device_or_resource_busy = EBUSY,
  directory_not_empty = ENOTEMPTY, executable_format_error = ENOEXEC, file_exists = EEXIST, file_too_large = EFBIG,
  filename_too_long = ENAMETOOLONG, function_not_supported = ENOSYS, host_unreachable = EHOSTUNREACH,
  identifier_removed = EIDRM, illegal_byte_sequence = EILSEQ, inappropriate_io_control_operation = ENOTTY,
#if defined(__FreeBSD__) && defined(EINTEGRITY) // FreeBSD extension
...
```

Does anybody understand clang-format sufficiently to know what I have to do to avoid this weirdness? :)

DimitryAndric commented:
It's very strange. If you use `#ifdef`, clang-format formats as you would expect it, i.e. one enum value per line. If you use `#if defined(FOO)` it suddenly starts to concatenate as many enum values on one line as fit in 120 chars. Maybe we should disable clang-format for this particular piece of code?

philnik777 commented:
Yeah, I think you can just disable clang-format here. My guess is also that clang-format doesn't recognize it as an enum anymore. We should also document the linux-specific flags and guard them behind a `defined(__linux__)`, but that's probably out-of-scope for this patch. It would be good to file a bug report though.

DimitryAndric committed:
Turn off clang-format for `enum class errc`, as it gets confused by the
`_LIBCPP_DECLARE_STRONG_ENUM` macro. Also, go back to the nicely aligned
errno values, and indent the synopsis part similar to the actual code.

DimitryAndric committed:
Minor nits.

DimitryAndric commented:
> We should also document the linux-specific flags and guard them behind a `defined(__linux__)`, but that's probably out-of-scope for this patch. It would be good to file a bug report though.

FWIW, I looked at the whole list of errc values, and all of them except `EINTEGRITY` are defined in https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4849.pdf#subsection.19.5.1 . So I guess if there are any Linux specific ones they have crept into the standard itself, and we don't have to bother with `#ifdef`s. :)

philnik777 commented:
I couldn't find some of them in the latest draft. They have been deprecated in C++23 through https://wg21.link/LWG3869. Given that they are reserved by the standard, we indeed don't have to check them only on specific platforms. I guess we'll make them extensions at some point.

DimitryAndric commented:
I'll sit on this a bit longer, until more people return from holidays etc.

ldionne reviewed (commented):
According to its documentation, `std::errc` is intended to provide a *portable* way of referring to POSIX error codes. This extension goes against that ideal, since if you ever use `integrity_check_failed` in your code, it became non-portable (not only will it work only on FreeBSD, but it will also only work with libc++ on FreeBSD). That seems like a non-ideal situation, and I wonder whether it's really a good idea to provide that extension at all.

If someone writes code that uses `EINTEGRITY`, I would expect to see it under some platform-specific `#ifdef` and so it is probably acceptable for such code to look a bit uglier, e.g. `static_cast<int>(errc) == EINTEGRITY` instead of `errc == std::errc::integrity_check_failed`.

So overall, I am uncertain of the desirability of this extension, but of course I am happy to have my mind changed.

DimitryAndric commented:
I think errno values are inherently system-dependent, and it is mostly serendipity that a bunch of POSIX or Linux specific values ended up in the "blessed" list from the committee. If they are also starting do deprecate a bunch of these in the future, the utility value of the whole `errc` enum seems to be rather  in decline, and people will likely use other, even more non-portable methods to convey error codes.

In any case, I think we'll just keep on carrying that local patch for the time being. :)
